Re-wrote front-end scripting in TypeScript

ReidLitkauo opened the pull request:
It's a big commit, but there's no point in committing a half-done port.

ReidLitkauo committed:
Re-wrote front-end scripting in TypeScript

It's a big commit, but there's no point in committing a half-done port.

franaiello commented on src/_testing/UnitTestBase.ts:
```diff
@@ -0,0 +1,66 @@
+export default abstract class UnitTestBase {
```
Is the intention to keep your tests underneath '_testing'?  I would consider having the following in a separate PR where you separate your './src' from your './test' and the reason why is that you don't want your tests to be bundled with your production code partially for best practices but also for bloat.

./project
--src
---- employee.ts
--test
----employee.spec.ts

ReidLitkauo replied:
This shouldn't be hard to do. I'll make that separation.

franaiello commented on src/_common/script/AnimatedComponent/AnimatedComponent.ts:
```diff
@@ -0,0 +1,22 @@
+
+export default abstract class AnimatedComponent<Props, State> extends React.Component<Props, State> {
+
+	private animationIndex: number
```
I like the fact that you are considering 'private' vs 'public' or other...

franaiello commented on src/_common/script/AnimatedComponent/_ComponentAnimationManager.ts:
```diff
@@ -0,0 +1,47 @@
+import AnimationAttributes from "./_AnimationAttributes"
```
Why is this in  an underscore directory?

ReidLitkauo replied:
That `"./_AnimationAttributes"` is actually a file ... for some reason, TypeScript didn't like me adding the `.ts` extension when importing.

I decided that files with a leading underscore would be private dependencies that ought not be included by a file outside the folder they live in. So an external file could include `"../AnimatedComponent/AnimatedComponent"`, but not `"../AnimatedComponent/_ComponentAnimationManager"` or `"../AnimatedComponent/_AnimationAttributes" directly. I don't have any way of enforcing this though.

So this line is one private dependency including another private dependency. Neither file should be included by a file outside this directory.

franaiello commented on src/_common/script/AnimatedComponent/_ComponentAnimationManager.ts:
```diff
@@ -0,0 +1,47 @@
+
+	private sortMapByAnimationIndexAscending = () => {
+		this.mapOfComponentsToAnimationAttributes = new Map<React.RefObject<HTMLElement>, AnimationAttributes>([...this.mapOfComponentsToAnimationAttributes].sort((a: [React.RefObject<HTMLElement>, AnimationAttributes], b: [React.RefObject<HTMLElement>, AnimationAttributes]): number => {
+			return a[1].animationIndex > b[1].animationIndex ? 1 : 0
```
Why use an index hardcoded to '1'?  

When I look at this I am scratching my head what '1' means.  I guess my point is that you as the developer knows what '1' means but the reviewer may not - so I would consider something like below.

```
const TOP_ELEMENT = 1;
const BOTTOM_ELEMENT = 10;
const ANIMATION = 1;
const BLURRY = 2;
```

ReidLitkauo replied:
I've got the types for `a` and `b` written out on line 35, kind of near the end. `Array.prototype.sort()` zips up the key and value of the map elements I'm sorting, and it stores the key in position 0 and the value in position 1.

franaiello commented on src/_testing/UnitTestBase.ts:
```diff
@@ -0,0 +1,66 @@
+	//==========================================================================
+
+	private buildTestCases = () => {
+		for (let i = 0; i < this.numTestCases; i++) {
```
Given that this is zero based - should this read...

```
for (let i = 0; i <= this.numTestCases - 1; i++)
```

ReidLitkauo replied:
I think your line and my line are equivalent.

franaiello commented on src/_testing/UnitTestBase.ts:
```diff
@@ -0,0 +1,66 @@
+		expect(zippedActual).toStrictEqual(zippedExpected)
+	}
+
+	private zipTestCasesAndResults = (testCases: any[], results: any[]): any[] => {
```
Why are you doing this function?

ReidLitkauo replied:
I'm using Jest's `expect().toStrictEqual()` method to test if two objects are equal. Zipping up the test cases with the expected and actual results allows me to see what inputs failed.

franaiello commented on src/_testing/UnitTestExample.test.ts:
```diff
@@ -0,0 +1,30 @@
+// Using this temporary function in place of a testable module
+let sum = (a: number, b: number): number => a + b
+
+export default class UnitTestExample extends UnitTestBase {
```
Not a big fan of unit test files that 'extend' - personally unit tests should stand on their own at least in my rulebook

ReidLitkauo replied:
I built a bit of a unit test framework myself that sits on top of Jest. All my unit tests extend from either `UnitTestBase` or `UnitTestByExamplesBase`, which contain all the common testing boilerplate code. This lets me write new unit tests by simply building cases and running them through the function I'm trying to test.

franaiello commented on src/index/script/CanvasPosition/CanvasPosition.ts:
```diff
@@ -0,0 +1,165 @@
+	//--------------------------------------------------------------------------
+
+	private setRawCoords = (pos: Point) => {
+		this.raw.pos = pos
```
Only one line in a function - are you expecting more code here at some point in the future?

ReidLitkauo replied:
Probably not. I just liked how nice it made `setWithNormalizedCoords` look. I figured it was worth the additional verbosity to keep `setWithNormalizedCoords` at a consistent level of abstraction.

franaiello commented on src/index/script/CanvasPosition/_CanvasPosition_addPos.test.ts:
```diff
@@ -0,0 +1,29 @@
+
+export default class CanvasPosition_addPos extends CanvasPosition_setPos {
+	
+	protected buildOneTestCaseAndExpected(i: number): [any, any] {
```
I would encourage you to use more descriptive names for your tests such as 'validateTwoPointsShouldEqualEachOther' or something that another dev can look at and go 'ahh' I get the intent of this test case.  I know ugly name - but it does communicate a message to another dev.

ReidLitkauo replied:
I think a re-visit of how I do testing is in order. I probably need a `UnitTestByScattershot` base class, since that's how most of my tests are run; I genereate a ton of random expected outputs, derive corresponding inputs, then fire those derived inputs through the testing function. The base class a unit test derives from should tell the reader what type of test they're looking at.

franaiello commented on src/index/script/CanvasPosition/_CanvasPosition_setZoomLevel.test.ts:
```diff
@@ -0,0 +1,61 @@
+		]
+	}
+
+	//==========================================================================
```
Why are these lines in there?

franaiello commented on src/index/script/HeadsUpDisplay/PositionWidget/_PositionWidgetState.ts:
```diff
@@ -0,0 +1,6 @@
+
+
+export default class PositionWidgetState {
+
```
Is this meant to be blank?  If so - add a TODO comment or similar to communicate intent... but normally this should be ripped out.

ReidLitkauo replied:
It is intended to be blank -- TypeScript wants a state and props when extending `React.Component`, so this is here just to satisfy that requirement. I might fill it in later if I need to. I'll make a comment explaining this.

franaiello reviewed (changes requested):
I would also like to see a README.md at some point so if I were to pull this code down it would instruct me how to 'build' and then run unit tests without me having to engage you.

ReidLitkauo committed:
Cleaned up testing

Separated all tests into their own directory. Expanded the testing framework a bit to make it make more sense.
